tests/unittest_nnstreamer-edge.cc

```diff
@@ -47,6 +123,150 @@ _test_edge_event_cb (nns_edge_event_h event_h, void *user_data)
   return NNS_EDGE_ERROR_NONE;
 }
 
+/**
+ * @brief Edge thread for test.
+ */
+static void *
+_test_edge_thread (void *data)
+{
+  ne_test_data_s *_td = (ne_test_data_s *) data;
+  int ret;
+
+  ret = nns_edge_start (_td->handle, _td->is_server);
+  EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+  _td->running = true;
+  g_main_loop_run (_td->loop);
+  _td->running = false;
+
+  return NULL;
+}
+
+/**
+ * @brief Connect to local host, multiple clients.
+ */
+TEST(edge, connectLocal) {
+  nns_edge_h server_h, client1_h, client2_h;
+  ne_test_data_s *_td_server, *_td_client1, *_td_client2;
+  nns_edge_data_h data_h;
+  pthread_t server_thread, client1_thread, client2_thread;
+  pthread_attr_t attr;
+  size_t data_len;
+  void *data;
+  unsigned int i, retry;
+  int ret, port;
+  char *val;
+
+  _td_server = _get_test_data (true);
+  _td_client1 = _get_test_data (false);
+  _td_client2 = _get_test_data (false);
+  port = nns_edge_get_available_port ();
+
+  /* Prepare server (127.0.0.1:port) */
+  val = nns_edge_strdup_printf ("%d", port);
+  nns_edge_create_handle ("temp-server", "temp-topic", &server_h);
+  nns_edge_set_event_callback (server_h, _test_edge_event_cb, _td_server);
+  nns_edge_set_info (server_h, "IP", "127.0.0.1");
+  nns_edge_set_info (server_h, "PORT", val);
+  nns_edge_set_info (server_h, "CAPS", "test server");
+  _td_server->handle = server_h;
+  nns_edge_free (val);
+
+  /* Prepare client */
+  nns_edge_create_handle ("temp-client1", "temp-topic", &client1_h);
+  nns_edge_set_event_callback (client1_h, _test_edge_event_cb, _td_client1);
+  nns_edge_set_info (client1_h, "CAPS", "test client1");
+  _td_client1->handle = client1_h;
+
+  nns_edge_create_handle ("temp-client2", "temp-topic", &client2_h);
+  nns_edge_set_event_callback (client2_h, _test_edge_event_cb, _td_client2);
+  nns_edge_set_info (client2_h, "CAPS", "test client2");
+  _td_client2->handle = client2_h;
+
+  /* Start server/client thread */
+  pthread_attr_init (&attr);
+  pthread_attr_setdetachstate (&attr, PTHREAD_CREATE_DETACHED);

```

Comment:
Set the thread attribute as JOINABLE?

Reply:
IMO, unnecessary. this thread handles test data to quit g-main-loop.